x-pack/plugins/enterprise_search/SERVER.md

```diff
@@ -0,0 +1,88 @@
+- Lean on `elasticsearch-js` (the built-in Elasticsearch client) types where possible to do the heavy lifting.
+
+### Working with Elasticsearch and elasticsearch-js
+
+Any Elasticsearch function should be available via the client, but it can be a bit hard to figure out what the required function name is. Because the (elasticsearch-js)[elasticsearch-js] documentation isn't great, a search through Kibana's codebase is often more efficient than trying to go through the docs.
+
+The Kibana plugin provides an Elasticsearch client object to every route. You can access this client under the obvious name `client`. That client comes with two users you can use to perform actual Elasticsearch operations: `client.asCurrentUser` and `client.asInternalUser`. `asCurrentUser` will execute operations using the permissions attached to the incoming request, which should be the permissions of the user firing that request in Kibana. `asInternalUser` will execute requests as Kibana's internal user. This user has very limited permissions and is generally only useful if you want to manipulate Kibana's Saved Objects.
+
+For most operations you'll want to specify a generic type argument to indicate the expected return type, as TypeScript has no way of knowing that type. See:
+
+```
+const connectorResult = await client.asCurrentUser.search<ReturnType>({
+      from: accumulator.length,
+      index,
+      query,
+      size: 1000,
+    });
+```
+
+This will return a search object with each result document typed as the ReturnType specified in angle brankets.
+
+### Atomic updates
+
+When updating a document, you can use `client.asCurrentUser.update` to perform an atomic update.
+```typescript
+client.asCurrentUser.update({
+  doc: { property_to_update: 'new value' },
+  id: 'doc_id',
+  index: 'indexName',
+  })
+```
+ This will replace the specified properties with the values you provide, while leaving unspecified properties untouched.
+
+ If your requirements are more complicated, consider using [optimistic concurrency control](https://www.elastic.co/guide/en/elasticsearch/reference/current/optimistic-concurrency-control.html).
+
+
+### Pagination
+
+For consistency in pagination, we have a [Paginate<T>](common/types/pagination.ts) type that produces a paginated type, to be used in paginated results. This type works with Elastic EUI's paginated tables and provides a consistent interface for result types.
+
+For pagination inputs, take a look at [fetch_sync_jobs.ts](server/lib/connectors/fetch_sync_jobs.ts). Generally speaking you'll want to specify at minimum a `size` and a `page` index.
+ ### Testing
+
+ We should aim for 100% unit test coverage in the server, although you're allowed to deviate from that if the effort to get there doesn't make add much security. We have a longer-term roadmap item to add Kibana FTR configs for end-to-end tests so that we can run these against an actual Elasticsearch backend, but we don't have these yet.
```

Comment:
```suggestion

### Testing

We should aim for 100% unit test coverage in the server, although you're allowed to deviate from that if the effort to get there doesn't make add much security. We have a longer-term roadmap item to add Kibana FTR configs for end-to-end tests so that we can run these against an actual Elasticsearch backend, but we don't have these yet.
```